Add brief python az aro howto (#516)
Azure/ARO-RP: docs/az-aro-python-development.md

# How-to `az aro` development

ARO-RP distributed the CLI command `az aro` to control the cluster. 
Currently the `aro` subcommand is merged in the 
[azure-cli](https://github.com/azure/azure-cli) code base and 
is present as a command.

Currently, there are two codebases for `aro` subcommand:

* [upstream](https://github.com/Azure/ARO-RP/tree/master/python/az/aro) in main ARO-RP repo - which is an `az` extension
* [downstream][https://github.com/Azure/azure-cli/tree/dev/src/azure-cli/azure/cli/command_modules/aro] - which is an `az` command

Command is distributed alongside `az` and is automaticaly present in Azure cloud shell. Extension is installed as user action.

Currently code have to be in sync for upstream and downstream.

## Set up dev environment `az` for `aro` command

Python virtual environment is required:

```

> python3 -m venv env

Reviewer: FWIW the virtualenv this provides is not the same as the OS package virtualenv that we use elsewhere. There are some subtle differences.

source <az dir>/env/bin/activate
```

in the `az` project directory works the best in case of command development.

To work with `az` helper tool `azdev` comes in handy

```
pip install azdev
```

`azdev` tool **runs only** in the virtual environment.

Then run

```
azdev setup -c

> ```

Reviewer 2: Added comment. I think this is both. When working in `azure-cli` they recomend to use separete virtual env. I think it up to developer to decide configuration (use our viirtual env) or create new one.

> ```

Reviewer 3: ok, I'm looking for a default simple flow that makes it easier to onboard people.  If we can build this around something like `make pyenv`, that'd be good.

> ```

Author: For practical purposes, the `venv` module does the same as `virtualenv`. There are benefits to using it. It is part of core python, therefore it does not require an additional package. It is able to use python internals and does not have to do any tricks.

As far as I can understand the `make pyenv`, they can be synchronized.

I would do this in the follow-up PR.
However, this requires further thinking. Since there are now two flows

1) the code in ARO-RP, which is for the extension
2) the code in azure-cli, which is for the subcommand

They differ slightly, and the flow of work need to be aligned carefully to introduce the leat amount of confusion.

* [downstream][https://github.com/Azure/azure-cli/tree/dev/src/azure-cli/azure/cli/command_modules/aro] - which is an `az` command

Command is distributed alongside `az` and is automaticaly present in Azure cloud
shell. Extension is installed as user action.

Currently code have to be in sync for upstream and downstream.

## Set up dev environment `az` for `aro` command

Python virtual environment is required:

```
python3 -m venv env
source <az dir>/env/bin/activate
```

in the `az` project directory works the best in case of command development.

To work with `az` helper tool `azdev` comes in handy

```
pip install azdev
```

`azdev` tool **runs only** in the virtual environment.

Then run

```
azdev setup -c
```
if you are working on `aro extension`, you can use existing virtual env in
`ARO-RP` repository:

```
make pyenv
```

and follow the guide.

Once setup, the work on tool can be started.

## Code

`aro` subcommand/extension is placed in:

```
# azure-cli repository
.../azure-cli/src/azure-cli/azure/cli/command_modules/aro
# ARO-RP repository
.../ARO-RP/python/az/aro/azext_aro
```

with tests in corresponding repository `tests` folder.

> There is wery useful guide for authoring commands in the

Reviewer 4: ```suggestion
There is very useful guide for authoring commands in the
```

[azure-cli](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules) repository.

## Tests

Tests can be run with

```
azdev test aro (--live) (--lf) (--verbose) (--debug)
```

there are two main types of tests:

* live tests, that gets recorded and replayed
* mocks

The guide for writting and operating the tests is in the
[azure-cli](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_tests.md) repository.

### Live tests and recording

Tests can be recorded live, which means next run will be run against
the recorded values. For the recording python VCR library is used, and
recorded "tapes" are stored in the directory:

```
.../azure-cli/src/azure-cli/azure/cli/command_modules/aro/tests/latest/recordings
```

The test are recorded when the parameter `--live` is present, execution is
forced to perform against live Azure API. Every call is than recorded. Only failed
tests can be re-ran using `--lf` flag.

Recorded tests can be run as normal tests without the `--live`.

## Caveats

* `azure-cli` CI seems to not be isolated. Every run can fail or pass depending
on the state of `dev` branch. Force push to rerun the tests.

> * Rebase has the potential to butcher the `venv`. Delete and recreate it. This
> solves the issues.

Reviewer 4: ```suggestion
* Rebase has the potential to butcher the `venv`. To fix a broken `venv`, delete and recreate the `venv`.
```

* Rebases are a must as the dependencies are downloaded in wrong versions and even

> local environment tend to break against the API.

Reviewer 4: this is not very clear..
